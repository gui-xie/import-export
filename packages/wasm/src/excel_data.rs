use wasm_bindgen::prelude::*;

const ROOT_DATA_KEY: &str = "root";

#[wasm_bindgen(getter_with_clone)]
#[derive(Clone, Debug)]
pub struct ExcelColumnData {
    pub key: String,
    pub value: String,
    pub children: Vec<ExcelRowData>,
}

impl ExcelColumnData {
    pub fn new<T: Into<String>>(key: T, value: T) -> Self {
        ExcelColumnData {
            key: key.into(),
            value: value.into(),
            children: Vec::new(),
        }
    }

    pub fn new_group(group_name: String, value: String, children: Vec<ExcelRowData>) -> Self {
        ExcelColumnData {
            key: group_name.into(),
            value: value.into(),
            children,
        }
    }

    pub fn new_root_group(group_name: String, children: Vec<ExcelRowData>) -> Self {
        ExcelColumnData::new_group(group_name, "".into(), children)
    }

    pub fn get_children_len(&self) -> usize {
        let mut result = 0;
        for row in self.children.iter() {
            let mut row_len = 1;
            for column in row.columns.iter() {
                row_len += column.get_children_len();
            }
            result += row_len;
        }
        result
    }
}

#[wasm_bindgen]
impl ExcelColumnData {
    #[wasm_bindgen(constructor)]
    pub fn bind_new(key: String, value: String) -> ExcelColumnData {
        ExcelColumnData::new(key, value)
    }

    #[wasm_bindgen(js_name = newRoot)]
    pub fn new_root(children: Vec<ExcelRowData>) -> ExcelColumnData {
        ExcelColumnData {
            key: ROOT_DATA_KEY.into(),
            value: "".into(),
            children,
        }
    }

    #[wasm_bindgen(js_name=withChildren)]
    pub fn with_children(mut self, children: Vec<ExcelRowData>) -> Self {
        self.children = children;
        self
    }
}

#[wasm_bindgen(getter_with_clone)]
#[derive(Clone, Debug)]
pub struct ExcelRowData {
    pub columns: Vec<ExcelColumnData>,
}

impl ExcelRowData {
    pub fn get_row_len(&self) -> u32 {
        let mut row_len = 1;
        for (_, column_data) in self.columns.iter().enumerate() {
            let len = column_data.get_children_len();
            if len > row_len {
                row_len += len;
            }
        }
        row_len as u32
    }
}

#[wasm_bindgen]
impl ExcelRowData {
    #[wasm_bindgen(constructor)]
    pub fn new(columns: Vec<ExcelColumnData>) -> ExcelRowData {
        ExcelRowData { columns }
    }
}

#[wasm_bindgen(getter_with_clone)]
#[derive(Clone, Debug)]
pub struct ExcelData {
    pub rows: Vec<ExcelRowData>,
}

#[wasm_bindgen]
impl ExcelData {
    #[wasm_bindgen(constructor)]
    pub fn new(rows: Vec<ExcelRowData>) -> ExcelData {
        ExcelData { rows }
    }
}
