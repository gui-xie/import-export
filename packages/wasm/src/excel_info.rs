use wasm_bindgen::prelude::*;

#[wasm_bindgen(getter_with_clone)]
pub struct ExcelInfo {
    pub name: String,
    pub sheet_name: String,
    pub columns: Vec<ExcelColumnInfo>,
    pub author: String,
    pub create_time: String,
    pub title: Option<String>,
    pub default_row_height: Option<f64>,
    pub title_height: Option<f64>,
    pub title_color: Option<String>,
    pub title_text_color: Option<String>,
    pub title_text_bold: Option<bool>,
}

#[wasm_bindgen]
impl ExcelInfo {
    #[wasm_bindgen(constructor)]
    pub fn new(
        name: String,
        sheet_name: String,
        columns: Vec<ExcelColumnInfo>,
        author: String,
        create_time: String,
    ) -> ExcelInfo {
        ExcelInfo {
            name,
            sheet_name,
            columns,
            author,
            create_time,
            title: None,
            default_row_height: None,
            title_height: None,
            title_color: None,
            title_text_color: None,
            title_text_bold: None,
        }
    }

    #[wasm_bindgen(js_name = withTitle)]
    pub fn with_title(mut self, title: String) -> Self {
        self.title = Some(title);
        self
    }

    #[wasm_bindgen(js_name = withDefaultRowHeight)]
    pub fn with_default_row_height(mut self, row_height: f64) -> Self {
        self.default_row_height = Some(row_height);
        self
    }

    #[wasm_bindgen(js_name = withTitleHeight)]
    pub fn with_title_height(mut self, title_height: f64) -> Self {
        self.title_height = Some(title_height);
        self
    }

    #[wasm_bindgen(js_name = withTitleColor)]
    pub fn with_title_color(mut self, title_color: String) -> Self {
        self.title_color = Some(title_color);
        self
    }

    #[wasm_bindgen(js_name = withTitleTextColor)]
    pub fn with_title_text_color(mut self, title_text_color: String) -> Self {
        self.title_text_color = Some(title_text_color);
        self
    }

    #[wasm_bindgen(js_name = withTitleTextBold)]
    pub fn with_title_text_bold(mut self, title_text_bold: bool) -> Self {
        self.title_text_bold = Some(title_text_bold);
        self
    }

    pub fn get_header_row_index(&self) -> usize {
        if self.title.is_some() {
            1
        } else {
            0
        }
    }
}

#[wasm_bindgen(getter_with_clone)]
#[derive(Clone)]
pub struct ExcelColumnInfo {
    pub key: String,
    pub name: String,
    pub width: Option<f64>,
    pub note: Option<String>,
    pub data_type: ExcelDataType,
    pub allowed_values: Option<Vec<String>>,
    pub color: Option<String>,
    pub text_color: Option<String>,
    pub text_bold: bool,
}

#[wasm_bindgen]
#[derive(Clone, PartialEq)]
pub enum ExcelDataType {
    Text,
    Number,
}

#[wasm_bindgen]
impl ExcelColumnInfo {
    #[wasm_bindgen(constructor)]
    pub fn new(key: String, name: String, width: Option<f64>) -> ExcelColumnInfo {
        ExcelColumnInfo {
            key,
            name,
            width,
            note: None,
            data_type: ExcelDataType::Text,
            allowed_values: None,
            color: None,
            text_color: None,
            text_bold: true,
        }
    }

    #[wasm_bindgen(js_name = withNote)]
    pub fn with_note(mut self, note: String) -> Self {
        self.note = Some(note);
        self
    }

    #[wasm_bindgen(js_name = withDataType)]
    pub fn with_data_type(mut self, data_type: ExcelDataType) -> Self {
        self.data_type = data_type;
        self
    }

    #[wasm_bindgen(js_name = withAllowedValues)]
    pub fn with_allowed_values(mut self, allowed_values: Vec<String>) -> Self {
        self.allowed_values = Some(allowed_values);
        self
    }

    #[wasm_bindgen(js_name = withWidth)]
    pub fn with_width(mut self, width: f64) -> Self {
        self.width = Some(width);
        self
    }

    #[wasm_bindgen(js_name = withColor)]
    pub fn with_color(mut self, color: String) -> Self {
        self.color = Some(color);
        self
    }

    #[wasm_bindgen(js_name = withTextColor)]
    pub fn with_text_color(mut self, text_color: String) -> Self {
        self.text_color = Some(text_color);
        self
    }

    #[wasm_bindgen(js_name = withTextBold)]
    pub fn with_text_bold(mut self, text_bold: bool) -> Self {
        self.text_bold = text_bold;
        self
    }

    pub fn get_data_type(&self) -> String {
        match self.data_type {
            ExcelDataType::Text => "text".to_string(),
            ExcelDataType::Number => "number".to_string(),
        }
    }
}
