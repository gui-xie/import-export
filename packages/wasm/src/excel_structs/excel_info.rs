use std::collections::{HashMap, HashSet};

use wasm_bindgen::prelude::*;

#[wasm_bindgen(getter_with_clone)]
pub struct ExcelInfo {
    pub name: String,
    pub sheet_name: String,
    pub columns: Vec<ExcelColumnInfo>,
    pub author: String,
    pub create_time: String,
    pub title: Option<String>,
    pub title_format: Option<ExcelCellFormat>,
    pub title_height: Option<f64>,
    pub default_row_height: Option<f64>,
    pub dx: u16,
    pub dy: u32,
    pub is_header_freeze: bool,
}

impl ExcelInfo {
    pub fn new<T: Into<String>>(
        name: T,
        sheet_name: T,
        columns: Vec<ExcelColumnInfo>,
        author: T,
        create_time: T,
    ) -> Self {
        ExcelInfo::check_columns(&columns);
        ExcelInfo {
            name: name.into(),
            sheet_name: sheet_name.into(),
            columns,
            author: author.into(),
            create_time: create_time.into(),
            title: None,
            default_row_height: None,
            title_height: None,
            title_format: None,
            dx: 0,
            dy: 0,
            is_header_freeze: false,
        }
    }

    pub fn with_title<T: Into<String>>(mut self, title: T) -> Self {
        self.title = Some(title.into());
        self
    }

    pub fn get_parent_map(&self) -> HashMap<String, String> {
        let mut parent_keys = HashSet::new();
        for column in self.columns.iter() {
            if column.has_parent() {
                parent_keys.insert(column.parent.clone());
            }
        }
        let mut parent_map = HashMap::new();
        for column in self.columns.iter() {
            if parent_keys.contains(&column.key) {
                parent_map.insert(column.key.clone(), column.parent.clone());
            }
        }
        parent_map
    }

    pub fn get_leaf_columns(&self) -> Vec<&ExcelColumnInfo> {
        let parent_map = self.get_parent_map();
        let mut leaf_columns = Vec::new();
        for column in self.columns.iter() {
            if !parent_map.contains_key(&column.key) {
                leaf_columns.push(column);
            }
        }
        leaf_columns
    }

    fn check_columns(columns: &Vec<ExcelColumnInfo>) {
        let mut processed_columns = Vec::new();
        for column in columns.iter() {
            if column.has_parent() {
                let mut flag = false;
                while let Some(p) = processed_columns.pop() {
                    if p == column.parent {
                        flag = true;
                        processed_columns.push(p);
                        break;
                    }
                }
                if !flag {
                    panic!(
                        "Check Error: Parent column {} not found for column {}",
                        column.parent, column.key
                    );
                }
            }
            processed_columns.push(column.key.clone());
        }
    }
}

#[wasm_bindgen]
impl ExcelInfo {
    #[wasm_bindgen(constructor)]
    pub fn bind_new(
        name: String,
        sheet_name: String,
        columns: Vec<ExcelColumnInfo>,
        author: String,
        create_time: String,
    ) -> Self {
        ExcelInfo::new(name, sheet_name, columns, author, create_time)
    }

    #[wasm_bindgen(js_name = withTitle)]
    pub fn bind_with_title(self, title: String) -> Self {
        self.with_title(title)
    }

    #[wasm_bindgen(js_name = withDefaultRowHeight)]
    pub fn with_default_row_height(mut self, row_height: f64) -> Self {
        self.default_row_height = Some(row_height);
        self
    }

    #[wasm_bindgen(js_name = withTitleHeight)]
    pub fn with_title_height(mut self, title_height: f64) -> Self {
        self.title_height = Some(title_height);
        self
    }

    #[wasm_bindgen(js_name= withTitleFormat)]
    pub fn with_title_format(mut self, title_format: ExcelCellFormat) -> Self {
        self.title_format = Some(title_format);
        self
    }

    #[wasm_bindgen(js_name = withOffset)]
    pub fn with_offset(mut self, dx: u16, dy: u32) -> Self {
        self.dx = dx;
        self.dy = dy;
        self
    }

    #[wasm_bindgen(js_name = withIsHeaderFreeze)]
    pub fn with_is_header_freeze(mut self, is_header_freeze: bool) -> Self {
        self.is_header_freeze = is_header_freeze;
        self
    }
}

#[wasm_bindgen(getter_with_clone)]
#[derive(Clone)]
pub struct ExcelCellFormat {
    pub rule: String,
    pub value: String,
    pub color: String,
    pub bold: bool,
    pub italic: bool,
    pub underline: bool,
    pub strikethrough: bool,
    pub font_size: f64,
    pub background_color: String,
    pub align: String,
    pub align_vertical: String,
    pub date_format: Option<String>,
    pub border_color: Option<String>,
}

#[wasm_bindgen]
impl ExcelCellFormat {
    #[wasm_bindgen(constructor)]
    pub fn new() -> ExcelCellFormat {
        ExcelCellFormat {
            rule: "default".into(),
            value: "".into(),
            color: "black".into(),
            bold: false,
            italic: false,
            underline: false,
            strikethrough: false,
            font_size: 11.0,
            background_color: "white".into(),
            align: "left".into(),
            align_vertical: "bottom".into(),
            date_format: None,
            border_color: None,
        }
    }

    #[wasm_bindgen(js_name = withRule)]
    pub fn with_rule(mut self, rule: String) -> Self {
        self.rule = rule;
        self
    }

    #[wasm_bindgen(js_name = withValue)]
    pub fn with_value(mut self, value: String) -> Self {
        self.value = value;
        self
    }

    #[wasm_bindgen(js_name = withColor)]
    pub fn bind_with_color(self, color: String) -> Self {
        self.with_color(color)
    }

    #[wasm_bindgen(js_name = withBold)]
    pub fn with_bold(mut self, bold: bool) -> Self {
        self.bold = bold;
        self
    }

    #[wasm_bindgen(js_name = withItalic)]
    pub fn with_italic(mut self, italic: bool) -> Self {
        self.italic = italic;
        self
    }

    #[wasm_bindgen(js_name = withUnderline)]
    pub fn with_underline(mut self, underline: bool) -> Self {
        self.underline = underline;
        self
    }

    #[wasm_bindgen(js_name = withStrikethrough)]
    pub fn with_strikethrough(mut self, strikethrough: bool) -> Self {
        self.strikethrough = strikethrough;
        self
    }

    #[wasm_bindgen(js_name = withFontSize)]
    pub fn with_font_size(mut self, font_size: f64) -> Self {
        self.font_size = font_size;
        self
    }

    #[wasm_bindgen(js_name = withBackgroundColor)]
    pub fn with_background_color(mut self, background_color: String) -> Self {
        self.background_color = background_color;
        self
    }

    #[wasm_bindgen(js_name = withAlign)]
    pub fn bind_with_align(self, align: String) -> Self {
        self.with_align(align)
    }

    #[wasm_bindgen(js_name = withAlignVertical)]
    pub fn bind_with_align_vertical(self, align_vertical: String) -> Self {
        self.with_align_vertical(align_vertical)
    }

    #[wasm_bindgen(js_name = withDateFormat)]
    pub fn with_date_format(mut self, date_format: String) -> Self {
        self.date_format = Some(date_format);
        self
    }

    #[wasm_bindgen(js_name = withBorderColor)]
    pub fn bind_with_border_color(self, border_color: String) -> Self {
        self.with_border_color(border_color)
    }
}

impl ExcelCellFormat {
    pub fn with_color<T: Into<String>>(mut self, color: T) -> Self {
        self.color = color.into();
        self
    }

    pub fn with_align<T: Into<String>>(mut self, align: T) -> Self {
        self.align = align.into();
        self
    }

    pub fn with_align_vertical<T: Into<String>>(mut self, align_vertical: T) -> Self {
        self.align_vertical = align_vertical.into();
        self
    }

    pub fn with_border_color<T: Into<String>>(mut self, border_color: T) -> Self {
        self.border_color = Some(border_color.into());
        self
    }
}

#[wasm_bindgen(getter_with_clone)]
#[derive(Clone)]
pub struct ExcelColumnInfo {
    pub key: String,
    pub name: String,
    pub width: f64,
    pub note: Option<String>,
    pub data_type: String,
    pub allowed_values: Vec<String>,
    pub parent: String,
    pub format: Option<ExcelCellFormat>,
    pub value_format: Vec<ExcelCellFormat>,
    pub data_group: String,
    pub data_group_parent: String,
}

#[wasm_bindgen]
impl ExcelColumnInfo {
    #[wasm_bindgen(constructor)]
    pub fn bind_new(key: String, name: String) -> ExcelColumnInfo {
        ExcelColumnInfo::new(key, name)
    }

    #[wasm_bindgen(js_name = withNote)]
    pub fn with_note(mut self, note: String) -> Self {
        self.note = Some(note);
        self
    }

    #[wasm_bindgen(js_name = withDataType)]
    pub fn bind_with_data_type(self, data_type: String) -> Self {
        self.with_data_type(data_type)
    }

    #[wasm_bindgen(js_name = withAllowedValues)]
    pub fn with_allowed_values(mut self, allowed_values: Vec<String>) -> Self {
        self.allowed_values = allowed_values;
        self
    }

    #[wasm_bindgen(js_name = withWidth)]
    pub fn with_width(mut self, width: f64) -> Self {
        self.width = width;
        self
    }

    #[wasm_bindgen(js_name = withFormat)]
    pub fn with_format(mut self, format: ExcelCellFormat) -> Self {
        self.format = Some(format);
        self
    }

    #[wasm_bindgen(js_name = withParent)]
    pub fn bind_with_parent(self, parent: String) -> Self {
        self.with_parent(parent)
    }

    #[wasm_bindgen(js_name = withValueFormat)]
    pub fn with_value_format(mut self, value_format: Vec<ExcelCellFormat>) -> Self {
        self.value_format = value_format;
        self
    }

    #[wasm_bindgen(js_name = withDataGroup)]
    pub fn bind_with_data_group(self, group: String) -> Self {
        self.with_data_group(group)
    }

    #[wasm_bindgen(js_name = withDataGroupParent)]
    pub fn bind_with_data_group_parent(self, group_parent: String) -> Self {
        self.with_data_group_parent(group_parent)
    }
}

impl ExcelColumnInfo {
    pub fn new<T: Into<String>>(key: T, name: T) -> Self {
        ExcelColumnInfo {
            key: key.into(),
            name: name.into(),
            width: 10.0,
            note: None,
            data_type: "text".into(),
            allowed_values: Vec::new(),
            parent: "".into(),
            format: None,
            value_format: Vec::new(),
            data_group: "".into(),
            data_group_parent: "".into(),
        }
    }

    pub fn with_parent<T: Into<String>>(mut self, parent: T) -> Self {
        self.parent = parent.into();
        self
    }

    pub fn with_data_type<T: Into<String>>(mut self, data_type: T) -> Self {
        self.data_type = data_type.into();
        self
    }

    pub fn with_data_group<T: Into<String>>(mut self, data_group: T) -> Self {
        self.data_group = data_group.into();
        self
    }

    pub fn with_data_group_parent<T: Into<String>>(mut self, data_group_parent: T) -> Self {
        self.data_group_parent = data_group_parent.into();
        self
    }

    pub fn is_root_group(&self) -> bool {
        !self.data_group.is_empty() && self.data_group_parent.is_empty()
    }

    pub fn has_parent(&self) -> bool {
        !self.parent.is_empty()
    }

    pub fn get_value_format<'a>(&'a self, value: &'a String) -> Option<&'a ExcelCellFormat> {
        let mut result = None;
        for vf in self.value_format.iter() {
            if vf.rule == "eq" {
                if vf.value == *value {
                    result = Some(vf);
                    break;
                }
            }
            if vf.rule == "default" {
                result = Some(vf);
            }
        }
        result
    }
}
