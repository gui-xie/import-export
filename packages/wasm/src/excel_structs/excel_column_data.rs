use std::fmt;
use wasm_bindgen::prelude::*;
use super::excel_row_data::ExcelRowData;

#[wasm_bindgen(getter_with_clone)]
#[derive(Clone)]
pub struct ExcelColumnData {
    pub key: String,
    pub value: String,
    pub children: Vec<ExcelRowData>,
}

impl ExcelColumnData {
    pub fn new<T: Into<String>>(key: T, value: T) -> Self {
        ExcelColumnData {
            key: key.into(),
            value: value.into(),
            children: Vec::new(),
        }
    }

    pub fn get_children_len(&self) -> usize {
        let mut result = 0;
        for row in self.children.iter() {
            let mut row_len = 1;
            for column in row.columns.iter() {
                row_len += column.get_children_len();
            }
            result += row_len;
        }
        result
    }
}

#[wasm_bindgen]
impl ExcelColumnData {
    #[wasm_bindgen(constructor)]
    pub fn bind_new(key: String, value: String) -> ExcelColumnData {
        ExcelColumnData::new(key, value)
    }

    #[wasm_bindgen(js_name = newGroup)]
    pub fn new_group(group_name: String, value: String, children: Vec<ExcelRowData>) -> Self {
        ExcelColumnData {
            key: group_name.into(),
            value: value.into(),
            children,
        }
    }

    #[wasm_bindgen(js_name = newRootGroup)]
    pub fn new_root_group(group_name: String, children: Vec<ExcelRowData>) -> Self {
        ExcelColumnData::new_group(group_name, "".into(), children)
    }

    #[wasm_bindgen(js_name = withChildren)]
    pub fn with_children(mut self, children: Vec<ExcelRowData>) -> Self {
        self.children = children;
        self
    }
}

impl fmt::Debug for ExcelColumnData {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}: {}", self.key, self.value)?;
        Ok(())
    }
}
