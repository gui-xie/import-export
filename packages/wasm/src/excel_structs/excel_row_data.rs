use std::fmt;

use wasm_bindgen::prelude::*;

use super::excel_column_data::ExcelColumnData;

#[wasm_bindgen(getter_with_clone)]
#[derive(Clone)]
pub struct ExcelRowData {
    pub columns: Vec<ExcelColumnData>,
}

impl ExcelRowData {
    pub fn get_row_len(&self) -> u32 {
        let mut row_len = 1;
        for (_, column_data) in self.columns.iter().enumerate() {
            let len = column_data.get_children_len();
            if len > row_len {
                row_len += len;
            }
        }
        row_len as u32
    }
}

#[wasm_bindgen]
impl ExcelRowData {
    #[wasm_bindgen(constructor)]
    pub fn new(columns: Vec<ExcelColumnData>) -> ExcelRowData {
        ExcelRowData { columns }
    }
}

impl fmt::Debug for ExcelRowData {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for (_, column_data) in self.columns.iter().enumerate() {
            write!(f, "{:?},", column_data)?
        }
        Ok(())
    }
}
