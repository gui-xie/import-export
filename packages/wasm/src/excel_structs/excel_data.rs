use super::excel_row_data::ExcelRowData;
use std::fmt;
use wasm_bindgen::prelude::*;

#[wasm_bindgen(getter_with_clone)]
#[derive(Clone)]
pub struct ExcelData {
    pub rows: Vec<ExcelRowData>,
}

#[wasm_bindgen]
impl ExcelData {
    #[wasm_bindgen(constructor)]
    pub fn new(rows: Vec<ExcelRowData>) -> ExcelData {
        ExcelData { rows }
    }
}

impl fmt::Debug for ExcelData {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "data")?;
        for (_, column_data) in self.rows.iter().enumerate() {
            writeln!(f, "  {:?}", column_data)?
        }
        Ok(())
    }
}
