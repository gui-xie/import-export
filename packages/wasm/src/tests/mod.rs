#[cfg(test)]
mod tests {
    use crate::{create_template_buffer, excel_structs::*, export_data_buffer, import_data_buffer};

    use excel_column_data::*;
    use excel_data::*;
    use excel_info::*;
    use excel_row_data::*;
    use insta::{assert_binary_snapshot, assert_snapshot};
    use wasm_bindgen_test::*;

    wasm_bindgen_test_configure!(run_in_browser);

    fn create_excel_info() -> ExcelInfo {
        let name = "Pokemon";
        let sheet_name = "FireRed Pokédex";
        let author = "senlinz";
        let create_time = "2024-11-01T08:00:00";
        let columns = vec![
            ExcelColumnInfo::new("number", "Number"),
            ExcelColumnInfo::new("name", "Name"),
            ExcelColumnInfo::new("first_type", "First"),
            ExcelColumnInfo::new("second_type", "Second"),
            ExcelColumnInfo::new("abilities", "Abilities"),
            ExcelColumnInfo::new("hp", "HP")
                .with_data_type("number")
                .with_data_type("number"),
            ExcelColumnInfo::new("attack", "Attack")
                .with_data_type("number")
                .with_data_type("number"),
            ExcelColumnInfo::new("defense", "Defense")
                .with_data_type("number")
                .with_data_type("number"),
            ExcelColumnInfo::new("sp_attack", "Sp. Atk")
                .with_data_type("number")
                .with_data_type("number"),
            ExcelColumnInfo::new("sp_defense", "Sp. Def")
                .with_data_type("number")
                .with_data_type("number"),
            ExcelColumnInfo::new("speed", "Speed")
                .with_data_type("number")
                .with_data_type("number"),
            ExcelColumnInfo::new("total", "Total").with_data_type("number"),
        ];
        excel_info::ExcelInfo::new(name, sheet_name, columns, author, create_time)
    }

    #[test]
    fn import_pokemon_success() {
        // Arrange
        let info = create_excel_info();
        let excel_bytes: &[u8] = include_bytes!(
            "./snapshots/imexport_wasm__tests__tests__export_pokemon_success.snap.xlsx"
        );

        // Act
        let result = import_data_buffer(info, excel_bytes);

        // Assert
        assert!(result.is_ok());
        let result = result.unwrap();
        let snapshot = format!("{:?}", result);
        assert_snapshot!(snapshot);
    }

    #[test]
    fn create_pokemon_template_success() {
        // Arrange
        let info = create_excel_info();

        // Act
        let result = create_template_buffer(&info);

        // Assert
        assert!(result.is_ok());
        let result = result.unwrap();
        assert_binary_snapshot!("create_pokemon_template_success.xlsx", result);
    }

    #[test]
    fn export_pokemon_success() {
        // Arrange
        let info = create_excel_info();
        let data = excel_data::ExcelData {
            rows: vec![ExcelRowData::new(vec![
                ExcelColumnData::new("number", "#001"),
                ExcelColumnData::new("name", "Bulbasaur"),
                ExcelColumnData::new("first_type", "Grass"),
                ExcelColumnData::new("second_type", "Poison"),
                ExcelColumnData::new("abilities", "Overgrow/Chlorophyll"),
                ExcelColumnData::new("hp", "45"),
                ExcelColumnData::new("attack", "49"),
                ExcelColumnData::new("defense", "49"),
                ExcelColumnData::new("sp_attack", "65"),
                ExcelColumnData::new("sp_defense", "65"),
                ExcelColumnData::new("speed", "45"),
                ExcelColumnData::new("total", "318"),
            ])],
        };

        // Act
        let result = export_data_buffer(&info, &data);

        // Assert
        assert!(result.is_ok());
        let result = result.unwrap();
        assert_binary_snapshot!("export_pokemon_success.xlsx", result);
    }

    #[test]
    fn export_pokemon_skill_success() {
        // Arrange
        let info = ExcelInfo::new(
            "Pokemon",
            "FireRed Pokédex",
            vec![
                ExcelColumnInfo::new("number", "Number"),
                ExcelColumnInfo::new("name", "Name"),
                ExcelColumnInfo::new("moves", "Moves").with_data_group("moves"),
                ExcelColumnInfo::new("moves_by", "By")
                    .with_parent("moves")
                    .with_data_group("moves_by")
                    .with_data_group_parent("moves"),
                ExcelColumnInfo::new("moves_lv", "LV/TM/HM")
                    .with_parent("moves")
                    .with_data_type("number")
                    .with_data_group_parent("moves_by"),
                ExcelColumnInfo::new("moves_move", "Move.")
                    .with_parent("moves")
                    .with_width(16.0)
                    .with_data_group_parent("moves_by"),
                ExcelColumnInfo::new("moves_type", "Type.")
                    .with_parent("moves")
                    .with_data_group_parent("moves_by"),
                ExcelColumnInfo::new("moves_cat", "Cat.")
                    .with_parent("moves")
                    .with_data_group_parent("moves_by"),
                ExcelColumnInfo::new("moves_power", "Power.")
                    .with_parent("moves")
                    .with_data_group_parent("moves_by"),
                ExcelColumnInfo::new("moves_acc", "Acc.")
                    .with_parent("moves")
                    .with_data_group_parent("moves_by"),
            ],
            "senlinz",
            "2024-11-01T08:00:00",
        );
        let data = ExcelData::new(vec![ExcelRowData::new(vec![
            ExcelColumnData::new("number", "#001"),
            ExcelColumnData::new("name", "Bulbasaur"),
            ExcelColumnData::new_root_group(
                "moves".into(),
                vec![
                    ExcelRowData::new(vec![ExcelColumnData::new_group(
                        "moves_by".into(),
                        "Level".into(),
                        vec![
                            ExcelRowData::new(vec![
                                ExcelColumnData::new("moves_lv", "1"),
                                ExcelColumnData::new("moves_move", "Tackle"),
                                ExcelColumnData::new("moves_type", "Normal"),
                                ExcelColumnData::new("moves_cat", "Physical"),
                                ExcelColumnData::new("moves_power", "40"),
                                ExcelColumnData::new("moves_acc", "100"),
                            ]),
                            ExcelRowData::new(vec![
                                ExcelColumnData::new("moves_lv", "4"),
                                ExcelColumnData::new("moves_move", "Growl"),
                                ExcelColumnData::new("moves_type", "Normal"),
                                ExcelColumnData::new("moves_cat", "Status"),
                                ExcelColumnData::new("moves_power", "-"),
                                ExcelColumnData::new("moves_acc", "100"),
                            ]),
                            ExcelRowData::new(vec![
                                ExcelColumnData::new("moves_lv", "7"),
                                ExcelColumnData::new("moves_move", "Leech Seed"),
                                ExcelColumnData::new("moves_type", "Grass"),
                                ExcelColumnData::new("moves_cat", "Status"),
                                ExcelColumnData::new("moves_power", "-"),
                                ExcelColumnData::new("moves_acc", "90"),
                            ]),
                            ExcelRowData::new(vec![
                                ExcelColumnData::new("moves_lv", "10"),
                                ExcelColumnData::new("moves_move", "Vine Whip"),
                                ExcelColumnData::new("moves_type", "Grass"),
                                ExcelColumnData::new("moves_cat", "Special"),
                                ExcelColumnData::new("moves_power", "35"),
                                ExcelColumnData::new("moves_acc", "100"),
                            ]),
                        ],
                    )]),
                    ExcelRowData::new(vec![ExcelColumnData::new_group(
                        "moves_by".into(),
                        "Egg".into(),
                        vec![
                            ExcelRowData::new(vec![
                                ExcelColumnData::new("moves_move", "Curse"),
                                ExcelColumnData::new("moves_type", "Ghost"),
                                ExcelColumnData::new("moves_cat", "Status"),
                                ExcelColumnData::new("moves_power", "-"),
                                ExcelColumnData::new("moves_acc", "—"),
                            ]),
                            ExcelRowData::new(vec![
                                ExcelColumnData::new("moves_move", "Ingrain"),
                                ExcelColumnData::new("moves_type", "Grass"),
                                ExcelColumnData::new("moves_cat", "Status"),
                                ExcelColumnData::new("moves_power", "-"),
                                ExcelColumnData::new("moves_acc", "—"),
                            ]),
                            ExcelRowData::new(vec![
                                ExcelColumnData::new("moves_move", "Nature Power"),
                                ExcelColumnData::new("moves_type", "Normal"),
                                ExcelColumnData::new("moves_cat", "Status"),
                                ExcelColumnData::new("moves_power", "-"),
                                ExcelColumnData::new("moves_acc", "—"),
                            ]),
                        ],
                    )]),
                ],
            ),
        ])]);

        // Act
        let result = export_data_buffer(&info, &data);

        // Assert
        assert!(result.is_ok());
        let result = result.unwrap();
        assert_binary_snapshot!("export_pokemon_skill_success.xlsx", result);
    }
}
