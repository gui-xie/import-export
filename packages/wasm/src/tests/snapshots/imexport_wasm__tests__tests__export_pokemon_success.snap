---
source: src/tests/mod.rs
assertion_line: 106
expression: result
extension: xlsx
snapshot_kind: binary
---
